cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/MetricsMessageAggregator.java

```diff
@@ -0,0 +1,278 @@
+      }
+      byte[] payload = encodeOutputStream.toByteArray();
+      return (Message) new Message() {
+        @Override
+        public String getId() {
+          // TopicRelayers keep track of the last message ID
+          // to fetch messages in the next iteration.
+          // Message IDs are generated by the messaging service. Return the last message ID before aggregation
+          // every time.
+          return messageId[0];
+        }
+
+        @Override
+        public byte[] getPayload() {
+          return payload;
+        }
+      };
+    }).iterator();
+  }
+
+  private long getRoundedTimestamp(long timestamp, long offset) {
+    return (Math.max(timestamp - offset, 0) / aggregationWindowSeconds)
+        * aggregationWindowSeconds;
+  }
+
+  /**
+   * The key for aggregating {@link MetricValues}. Only values with the same key
+   * will be merged.
+   */
+  private static class MetricsMapKey {
+
+    private final Map<String, String> tags;
+    private final long roundedTimestamp;
+
+    private MetricsMapKey(Map<String, String> tags, long roundedTimestamp) {
+      this.tags = tags;
+      this.roundedTimestamp = roundedTimestamp;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (!(o instanceof MetricsMapKey)) {
+        return false;
+      }
+      MetricsMapKey that = (MetricsMapKey) o;
+      return roundedTimestamp == that.roundedTimestamp && tags.equals(
+          that.tags);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(tags, roundedTimestamp);
+    }
+
+    public Map<String, String> getTags() {
+      return tags;
+    }
+
+    public long getRoundedTimestamp() {
+      return roundedTimestamp;
+    }
+  }
+
+  /**
+   * Aggregator for {@link MetricValue}s. Users can add {@link MetricValue}s one
+   * by one and get the resulting {@link MetricValues} once all the metrics are
+   * added.
+   */
+  private static class MetricsAggregator {
+
+    private final Collection<MetricValue> metrics;
+    private final Map<String, AggregatedMetric> aggregatedMetricsMap;
+
+    private MetricsAggregator() {
+      this.metrics = new ArrayList<>();
+      this.aggregatedMetricsMap = new HashMap<>();
+    }
+
+    public void addMetric(MetricValue metric, long timestamp) {
+      switch (metric.getType()) {
+        case GAUGE:
+        case COUNTER:
+          aggregatedMetricsMap.computeIfAbsent(metric.getName(),
+              AggregatedMetric::new).addMetric(metric, timestamp);
+          break;
+        default:
+          // Don't know how to aggregate other metric types, so store them without aggregation.
+          metrics.add(metric);
+      }
+    }
+
+    public MetricValues getAggregateMetricValues(Map<String, String> tags,
+        long timestamp) {
+      metrics.addAll(
+          aggregatedMetricsMap.values().stream().map(AggregatedMetric::emit)
+              .collect(Collectors.toList()));
```

Comment:
No need to collect to a list and then add to another collection. Just do a `.forEach(metrics::add);` of the stream.